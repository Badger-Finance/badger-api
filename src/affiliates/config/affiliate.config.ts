import { BadRequest } from '@tsed/exceptions';
import { Chain } from '../../chains/config/chain.config';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';

type Affiliates = { [protocol: string]: Affiliate };

export abstract class Affiliate {
  private static affiliates: Affiliates = {};

  readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  static register(protocol: Protocol, affiliate: Affiliate): void {
    this.affiliates[protocol] = affiliate;
  }

  static getAffiliate(affiliateData: SettAffiliateData): Affiliate {
    const affiliate = this.affiliates[affiliateData.protocol];
    if (!affiliate) {
      throw new BadRequest(`${affiliate} is not a affiliate`);
    }
    return affiliate;
  }

  abstract getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData>;
}
