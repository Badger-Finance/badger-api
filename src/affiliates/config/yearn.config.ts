import { BigNumber, ethers } from 'ethers';
import { Chain } from '../../chains/config/chain.config';
import { guestListAbi } from '../../config/abi/guest-list.abi';
import { yearnAffiliateVaultWrapperAbi } from '../../config/abi/yearn-affiliate-vault-wrapper.abi';
import { Protocol } from '../../config/constants';
import { SettAffiliateData } from '../../setts/interfaces/sett-affiliate-data.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { getToken } from '../../tokens/tokens.utils';
import { Affiliate } from './affiliate.config';

export class Yearn extends Affiliate {
  constructor() {
    super(Protocol.Yearn);
    Affiliate.register(Protocol.Yearn, this);
  }

  async getAffiliateVaultData(chain: Chain, sett: SettDefinition): Promise<SettAffiliateData> {
    const affiliateData: SettAffiliateData = {
      protocol: Protocol.Yearn,
    };
    if (sett.hasBouncer) {
      try {
        const vaultWrapper = new ethers.Contract(sett.settToken, yearnAffiliateVaultWrapperAbi, chain.provider);
        const guestListAddress = await vaultWrapper.guestList();
        const guesList = new ethers.Contract(guestListAddress, guestListAbi, chain.provider);
        const [available, limit]: [BigNumber, BigNumber] = await Promise.all([
          guesList.remainingTotalDepositAllowed(),
          guesList.totalDepositCap(),
        ]);
        const depositToken = getToken(sett.depositToken);
        affiliateData.availableDepositLimit = parseFloat(ethers.utils.formatUnits(available, depositToken.decimals));
        affiliateData.depositLimit = parseFloat(ethers.utils.formatUnits(limit, depositToken.decimals));
      } catch {
        affiliateData.availableDepositLimit = 0;
        affiliateData.depositLimit = 50;
      }
    }
    return affiliateData;
  }
}
