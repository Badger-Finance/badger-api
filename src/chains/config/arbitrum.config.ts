import { Protocol } from '../../config/enums/protocol.enum';
import { SettState } from '../../config/enums/sett-state.enum';
import { Stage } from '../../config/enums/stage.enum';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { getCurveSettTokenBalance } from '../../protocols/strategies/convex.strategy';
import { BouncerType } from '../../rewards/enums/bouncer-type.enum';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { arbitrumTokensConfig } from '../../tokens/config/arbitrum-tokens.config';
import { ChainNetwork } from '../enums/chain-network.enum';
import { ArbitrumStrategy } from '../strategies/arbitrum.strategy';
import { Chain } from './chain.config';

export class Arbitrum extends Chain {
  constructor() {
    super(
      'Arbitrum',
      'arbitrum',
      '0xa4b1',
      ChainNetwork.Arbitrum,
      arbitrumTokensConfig,
      arbitrumSetts,
      rpc[ChainNetwork.Arbitrum],
      new ArbitrumStrategy(Object.keys(arbitrumTokensConfig)),
      2425847,
      '0x635EB2C39C75954bb53Ebc011BDC6AfAAcE115A6',
      '0x85E1cACAe9a63429394d68Db59E14af74143c61c',
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    const gasPrice = 2;
    return {
      rapid: gasPrice,
      fast: gasPrice,
      standard: gasPrice,
      slow: gasPrice,
    };
  }
}

export const arbitrumSetts: SettDefinition[] = [
  {
    name: 'Sushiswap Wrapped Ether/Sushi Helper',
    settToken: TOKENS.BARB_SUSHI_WETH_SUSHI,
    depositToken: TOKENS.ARB_SUSHI_WETH_SUSHI,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0x86f772C82914f5bFD168f99e208d0FC2C371e9C2',
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SUSHI_WETH_WBTC,
    depositToken: TOKENS.ARB_SUSHI_WETH_WBTC,
    createdBlock: 13163959,
    protocol: Protocol.Sushiswap,
    strategy: '0xA6827f0f14D0B83dB925B616d820434697328c22',
  },
  {
    name: 'Curve renBTC/wBTC',
    settToken: TOKENS.BARB_CRV_RENBTC,
    depositToken: TOKENS.ARB_CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto',
    settToken: TOKENS.BARB_CRV_TRICRYPTO,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13237551,
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve Tricrypto Light',
    settToken: TOKENS.BARB_CRV_TRICRYPTO_LITE,
    depositToken: TOKENS.ARB_CRV_TRICRYPTO,
    getTokenBalance: getCurveSettTokenBalance,
    createdBlock: 13321375,
    stage: Stage.Staging,
    state: SettState.Experimental,
    protocol: Protocol.Curve,
  },
  {
    name: 'Swapr Swapr/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_SWPR_WETH,
    depositToken: TOKENS.ARB_SWP_SWPR_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Wrapped BTC/Wrapped ETH',
    settToken: TOKENS.BARB_SWP_WBTC_WETH,
    depositToken: TOKENS.ARB_SWP_WBTC_WETH,
    createdBlock: 13315350,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
  {
    name: 'Swapr Badger/Wrapped ETH',
    bouncer: BouncerType.Badger,
    settToken: TOKENS.BARB_SWP_BADGER_WETH,
    depositToken: TOKENS.ARB_SWP_BADGER_WETH,
    createdBlock: 2188169,
    state: SettState.Guarded,
    protocol: Protocol.Swapr,
  },
];
