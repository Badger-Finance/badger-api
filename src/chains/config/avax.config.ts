import { Network, Protocol, VaultState, VaultVersion } from '@badger-dao/sdk';
import rpc from '../../config/rpc.config';
import { TOKENS } from '../../config/tokens.config';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { VaultDefinition } from '../../vaults/interfaces/vault-definition.interface';
import { Chain } from './chain.config';
import { BaseStrategy } from '../strategies/base.strategy';
import { avalancheTokensConfig } from '../../tokens/config/avax-tokens.config';
import { Stage } from '../../config/enums/stage.enum';

export class Avalanche extends Chain {
  constructor() {
    super(
      'Avalanche',
      'avalanche',
      '0xa86a',
      Network.Avalanche,
      avalancheTokensConfig,
      rpc[Network.Avalanche],
      avalancheSetts,
      new BaseStrategy(Network.Avalanche, Object.keys(avalancheTokensConfig)),
    );
    Chain.register(this.network, this);
  }

  async getGasPrices(): Promise<GasPrices> {
    return this.defaultGasPrice();
  }

  // TODO: Update if badger is ever updated
  getBadgerTokenAddress(): string {
    return TOKENS.BADGER;
  }
}

export const avalancheSetts: VaultDefinition[] = [
  {
    name: 'WBTC',
    vaultToken: TOKENS.BAVAX_WBTC,
    depositToken: TOKENS.AVAX_WBTC,
    protocol: Protocol.Aave,
    stage: Stage.Staging,
    state: VaultState.Experimental,
    version: VaultVersion.v1_5,
  },
];
