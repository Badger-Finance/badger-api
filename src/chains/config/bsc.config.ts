import { Network } from "@badger-dao/sdk";

import rpc from "../../config/rpc.config";
import { TOKENS } from "../../config/tokens.config";
import { bscTokensConfig } from "../../tokens/config/bsc-tokens.config";
import { BaseStrategy } from "../strategies/base.strategy";
import { Chain } from "./chain.config";

export class BinanceSmartChain extends Chain {
  constructor() {
    super(Network.BinanceSmartChain, bscTokensConfig, rpc[Network.BinanceSmartChain], new BaseStrategy(Network.BinanceSmartChain));
    Chain.register(this.network, this);
  }

  getBadgerTokenAddress(): string {
    return TOKENS.MULTI_BADGER;
  }
}
