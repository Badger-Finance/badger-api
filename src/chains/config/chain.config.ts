import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { STAGE } from '../../config/constants';
import { GasPrices } from '../../gas/interfaces/gas-prices.interface';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { TokenConfig } from '../../tokens/interfaces/token-config.interface';
import { ChainStrategy } from '../strategies/chain.strategy';
import { Network } from '@badger-dao/sdk';

type Chains = Record<string, Chain>;

export abstract class Chain {
  private static chains: Chains = {};
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly network: Network;
  readonly tokens: TokenConfig;
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
  readonly batchProvider: ethers.providers.JsonRpcBatchProvider;
  readonly strategy: ChainStrategy;
  readonly graphUrl: string;
  readonly blocksPerYear: number;
  readonly badgerTree?: string;
  readonly rewardsLogger?: string;

  constructor(
    name: string,
    symbol: string,
    chainId: string,
    network: Network,
    tokens: TokenConfig,
    setts: SettDefinition[],
    rpcUrl: string,
    strategy: ChainStrategy,
    blocksPerYear: number,
    badgerTree?: string,
    rewardsLogger?: string,
  ) {
    this.name = name;
    this.symbol = symbol;
    this.chainId = chainId;
    this.network = network;
    this.tokens = tokens;
    this.setts = setts.filter((sett) => !sett.stage || sett.stage === STAGE);
    this.provider = new ethers.providers.JsonRpcProvider(rpcUrl);
    this.batchProvider = new ethers.providers.JsonRpcBatchProvider(rpcUrl);
    this.strategy = strategy;
    this.graphUrl = `https://api.thegraph.com/subgraphs/name/badger-finance/badger-dao-setts${
      network !== Network.Ethereum ? `-${symbol.toLowerCase()}` : ''
    }`;
    this.blocksPerYear = blocksPerYear;
    this.badgerTree = badgerTree;
    this.rewardsLogger = rewardsLogger;
  }

  static register(network: Network, chain: Chain): void {
    Chain.chains[network] = chain;
    Chain.chains[chain.symbol] = chain;
    if (network === Network.Polygon) {
      Chain.chains['matic'] = chain;
    }
  }

  static getChain(network?: Network): Chain {
    if (!network) {
      network = Network.Ethereum;
    }
    const chain = this.chains[network];
    if (!chain) {
      throw new BadRequest(`${network} is not a supported chain`);
    }
    return chain;
  }

  abstract getGasPrices(): Promise<GasPrices>;
}
