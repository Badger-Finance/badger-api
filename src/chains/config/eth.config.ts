import { ethers } from 'ethers';
import { BADGER_URL, Provider, STRATEGIES, TOKENS } from '../../config/constants';
import { Protocol } from '../../config/enums/protocol.enum';
import { Stage } from '../../config/enums/stage.enum';
import { getCurveSettTokenBalance } from '../../indexer/strategies/convex.strategy';
import { SettDefinition } from '../../setts/interfaces/sett-definition.interface';
import { ethTokensConfig } from '../../tokens/config/eth-tokens.config';
import { getZsDiggTokenBalance } from '../../tokens/custom/zs-digg-balance';
import { ChainNetwork } from '../enums/chain-network.enum';
import { EthStrategy } from '../strategies/eth.strategy';
import { Chain } from './chain.config';

export class Ethereum extends Chain {
  constructor() {
    super(
      'Ethereum',
      'eth',
      '0x01',
      ChainNetwork.Ethereum,
      ethTokensConfig,
      ethSetts,
      new ethers.providers.JsonRpcProvider(Provider.Alchemy),
      new EthStrategy(),
      BADGER_URL,
      2425847,
    );
    Chain.register(ChainNetwork.Ethereum, this);
  }
}

export const ethSetts: SettDefinition[] = [
  {
    name: 'Curve.fi renBTC/wBTC/sBTC',
    createdBlock: 11380877,
    depositToken: TOKENS.CRV_SBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_SBTC,
    strategy: '0x8cbb86a7e0780a6fbefeec108f9b4b0aa8193e24',
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve.fi renBTC/wBTC',
    createdBlock: 11380872,
    depositToken: TOKENS.CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_RENBTC,
    strategy: '0xaa9b716ccd717761f40479cd81f8e3a5a7b4cad7',
    protocol: Protocol.Curve,
  },
  {
    name: 'Curve.fi tBTC/sBTC',
    createdBlock: 11380878,
    depositToken: TOKENS.CRV_TBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_TBTC,
    strategy: '0x1ac31c470b90e366c70efc1ac28d5d7fa2f1dbe1',
    protocol: Protocol.Curve,
  },
  // TODO: Remove once fully deprecated
  {
    name: 'Harvest renBTC/wBTC',
    createdBlock: 11380939,
    depositToken: TOKENS.CRV_RENBTC,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_HRENBTC,
    protocol: Protocol.Curve,
  },
  {
    name: 'Uniswap Wrapped BTC/Badger',
    createdBlock: 11380883,
    depositToken: TOKENS.UNI_BADGER_WBTC,
    settToken: TOKENS.BUNI_BADGER_WBTC,
    protocol: Protocol.Uniswap,
  },
  {
    name: 'Uniswap Wrapped BTC/Digg',
    createdBlock: 11680833,
    depositToken: TOKENS.UNI_DIGG_WBTC,
    settToken: TOKENS.BUNI_DIGG_WBTC,
    protocol: Protocol.Uniswap,
  },
  {
    name: 'Sushiswap Wrapped BTC/Wrapped Ether',
    createdBlock: 11539529,
    depositToken: TOKENS.SUSHI_ETH_WBTC,
    strategy: '0x7A56d65254705B4Def63c68488C0182968C452ce',
    settToken: TOKENS.BSUSHI_ETH_WBTC,
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Sushiswap Wrapped BTC/Badger',
    createdBlock: 11537600,
    depositToken: TOKENS.SUSHI_BADGER_WBTC,
    strategy: '0x3a494D79AA78118795daad8AeFF5825C6c8dF7F1',
    settToken: TOKENS.BSUSHI_BADGER_WBTC,
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Sushiswap Wrapped BTC/Digg',
    createdBlock: 11681030,
    depositToken: TOKENS.SUSHI_DIGG_WBTC,
    strategy: '0xaa8dddfe7DFA3C3269f1910d89E4413dD006D08a',
    settToken: TOKENS.BSUSHI_DIGG_WBTC,
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Digg',
    createdBlock: 11680422,
    depositToken: TOKENS.DIGG,
    settToken: TOKENS.BDIGG,
  },
  {
    name: 'Badger',
    createdBlock: 11380871,
    depositToken: TOKENS.BADGER,
    settToken: TOKENS.BBADGER,
  },
  {
    createdBlock: 12246710,
    name: 'Yearn Wrapped BTC',
    depositToken: TOKENS.WBTC,
    settToken: TOKENS.BVYWBTC,
  },
  {
    name: 'Sushiswap Wrapped BTC/ibBTC',
    createdBlock: 12383984,
    depositToken: TOKENS.SUSHI_IBBTC_WBTC,
    settToken: TOKENS.BSUSHI_IBBTC_WBTC,
    strategy: '0xf4146A176b09C664978e03d28d07Db4431525dAd',
    protocol: Protocol.Sushiswap,
  },
  {
    name: 'Digg Stabilization',
    balanceDecimals: 18,
    createdBlock: 12375910,
    depositToken: TOKENS.DIGG,
    experimental: true,
    getTokenBalance: getZsDiggTokenBalance,
    protocol: Protocol.Badger,
    settToken: TOKENS.BZS_DIGG,
    stage: Stage.Staging,
    strategy: STRATEGIES.BZS_DIGG,
  },
  {
    name: 'Convex hBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_HBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_HBTC,
    strategy: '0xff26f400e57bf726822eacbb64fa1c52f1f27988',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex pBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_PBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_PBTC,
    strategy: '0x1c1fd689103bbfd701b3b7d41a3807f12814033d',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex oBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_OBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_OBTC,
    strategy: '0x2bb864cdb4856ab2d148c5ca52dd7ccec126d138',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex bBTC',
    createdBlock: 12589485,
    depositToken: TOKENS.CRV_BBTC,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_BBTC,
    strategy: '0x4f3e7a4566320b2709fd1986f2e9f84053d3e2a0',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex Tricrypto',
    createdBlock: 12679976,
    depositToken: TOKENS.CRV_TRICRYPTO,
    experimental: true,
    getTokenBalance: getCurveSettTokenBalance,
    settToken: TOKENS.BCRV_TRICRYPTO,
    strategy: '0x05ec4356e1acd89cc2d16adc7415c8c95e736ac1',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex cvxCRV Helper',
    createdBlock: 12678302,
    depositToken: TOKENS.CVXCRV,
    experimental: true,
    settToken: TOKENS.BCVXCRV,
    strategy: '0x4FE836b28ADc7867290764885b24d3992dAeE819',
    protocol: Protocol.Convex,
  },
  {
    name: 'Convex CVX Helper',
    createdBlock: 12678302,
    depositToken: TOKENS.CVX,
    experimental: true,
    settToken: TOKENS.BCVX,
    strategy: '0x5e7a939f77E5c441e2Ba1Dc8Be5D42020193376D',
    protocol: Protocol.Convex,
  },
];
