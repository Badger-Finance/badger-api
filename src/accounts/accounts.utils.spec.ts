import { DataMapper } from '@aws/dynamodb-data-mapper';
import { GraphQLClient } from 'graphql-request';
import { Ethereum } from '../chains/config/eth.config';
import { TestStrategy } from '../chains/strategies/test.strategy';
import { TOKENS } from '../config/tokens.config';
import { UserQuery, UserSettBalance, UsersQuery } from '../graphql/generated/badger';
import { LeaderBoardType } from '../leaderboards/enums/leaderboard-type.enum';
import * as priceUtils from '../prices/prices.utils';
import { inCurrency } from '../prices/prices.utils';
import { VaultDefinition } from '../vaults/interfaces/vault-definition.interface';
import { getVaultDefinition } from '../vaults/vaults.utils';
import { defaultAccount, randomSnapshot, setupMapper, TEST_ADDR } from '../test/tests.utils';
import { getToken } from '../tokens/tokens.utils';
import {
  defaultBoost,
  getAccountMap,
  getAccounts,
  getCachedAccount,
  getCachedBoost,
  getUserAccount,
  toSettBalance,
} from './accounts.utils';

describe('accounts.utils', () => {
  const chain = new Ethereum();

  const testSettBalance = (vaultDefinition: VaultDefinition): UserSettBalance => {
    const settToken = getToken(vaultDefinition.vaultToken);
    const depositToken = getToken(vaultDefinition.depositToken);
    const toWei = (amt: number) => {
      const values = amt * Math.pow(10, settToken.decimals);
      return values.toString();
    };
    return {
      id: TEST_ADDR,
      netDeposit: 4,
      netShareDeposit: toWei(4),
      grossDeposit: 9,
      grossShareDeposit: toWei(9),
      grossWithdraw: 5,
      grossShareWithdraw: toWei(5),
      user: {
        id: TEST_ADDR,
        settBalances: [],
      },
      sett: {
        id: settToken.address,
        name: settToken.name,
        symbol: settToken.symbol,
        available: 1,
        pricePerFullShare: 1034039284374221,
        balance: 3,
        totalSupply: 5,
        netDeposit: 4,
        netShareDeposit: toWei(4),
        grossDeposit: 9,
        grossShareDeposit: toWei(9),
        grossWithdraw: 5,
        grossShareWithdraw: toWei(5),
        decimals: 18,
        token: {
          id: depositToken.address,
          name: depositToken.name,
          symbol: depositToken.symbol,
          decimals: depositToken.decimals,
          totalSupply: 21000000,
        },
        treeDistributions: [],
        harvests: [],
      },
    };
  };

  describe('getAccountMap', () => {
    describe('no saved account', () => {
      it('returns the account with the default account mapping', async () => {
        setupMapper([]);
        const actual = await getAccountMap([TEST_ADDR]);
        expect(actual).toMatchSnapshot();
      });
    });

    describe('a saved account', () => {
      it('returns the stored account', async () => {
        setupMapper([defaultAccount(TEST_ADDR), defaultAccount(TOKENS.BADGER), defaultAccount(TOKENS.DIGG)]);
        const actual = await getAccountMap([TEST_ADDR]);
        expect(actual).toMatchSnapshot();
      });
    });
  });

  describe('getCachedAccount', () => {
    describe('no saved account', () => {
      it('returns undefined', async () => {
        setupMapper([]);
        const actual = await getCachedAccount(TEST_ADDR);
        expect(actual).toMatchObject(defaultAccount(TEST_ADDR));
      });
    });

    describe('encounters an errors', () => {
      it('returns undefined', async () => {
        jest.spyOn(DataMapper.prototype, 'query').mockImplementation(() => {
          throw new Error();
        });
        const actual = await getCachedAccount(TEST_ADDR);
        expect(actual).toMatchObject(defaultAccount(TEST_ADDR));
      });
    });

    describe('a saved account', () => {
      it('returns the stored account', async () => {
        const expected = { address: TEST_ADDR, claimableBalances: [] };
        setupMapper([expected]);
        const actual = await getCachedAccount(TEST_ADDR);
        expect(actual).toMatchObject(expected);
      });
    });
  });

  describe('getUserAccount', () => {
    describe('user exists', () => {
      it('returns the user account', async () => {
        const sett = getVaultDefinition(new Ethereum(), TOKENS.BBADGER);
        const result: UserQuery = {
          user: {
            settBalances: [testSettBalance(sett)],
          },
        };
        jest.spyOn(GraphQLClient.prototype, 'request').mockImplementationOnce(async () => Promise.resolve(result));
        const user = await getUserAccount(new Ethereum(), TEST_ADDR);
        expect(user).toMatchObject(result);
      });
    });

    describe('user does not exist', () => {
      it('returns null', async () => {
        const result: UserQuery = {
          user: null,
        };
        jest.spyOn(GraphQLClient.prototype, 'request').mockImplementationOnce(async () => Promise.resolve(result));
        const user = await getUserAccount(new Ethereum(), TEST_ADDR);
        expect(user).toMatchObject(result);
      });
    });
  });

  describe('getAccounts', () => {
    describe('users exist', () => {
      it('returns a list of user accounts', async () => {
        const result: UsersQuery = {
          users: Object.values(TOKENS).map((token) => ({ id: token, settBalances: [] })),
        };
        let responded = false;
        jest.spyOn(GraphQLClient.prototype, 'request').mockImplementation(async () => {
          if (responded) {
            return { users: [] };
          }
          responded = true;
          return Promise.resolve(result);
        });
        const users = await getAccounts(new Ethereum());
        expect(users).toMatchObject(Object.values(TOKENS));
      });
    });

    describe('users do not exist', () => {
      it('returns an empty list', async () => {
        jest.spyOn(GraphQLClient.prototype, 'request').mockImplementationOnce(async () => Promise.resolve(null));
        const nullReturn = await getAccounts(new Ethereum());
        expect(nullReturn).toMatchObject([]);

        jest
          .spyOn(GraphQLClient.prototype, 'request')
          .mockImplementationOnce(async () => Promise.resolve({ users: null }));
        const nullUsers = await getAccounts(new Ethereum());
        expect(nullUsers).toMatchObject([]);
      });
    });
  });

  describe('toSettBalance', () => {
    const chain = new Ethereum();
    const strategy = new TestStrategy();

    const testToSettBalance = (settAddress: string) => {
      it.each([
        [undefined, 'usd'],
        ['eth', 'eth'],
        ['usd', 'usd'],
      ])('returns sett balance request in %s currency with %s denominated value', async (currency, _res) => {
        jest.spyOn(priceUtils, 'getPrice').mockImplementation(async (contract) => ({
          ...(await strategy.getPrice(contract)),
          updatedAt: Date.now(),
        }));
        const sett = getVaultDefinition(chain, settAddress);
        const snapshot = randomSnapshot(sett);
        setupMapper([snapshot]);
        const depositToken = getToken(sett.depositToken);
        const mockBalance = testSettBalance(sett);
        const actual = await toSettBalance(chain, mockBalance, currency);
        expect(actual).toBeTruthy();
        expect(actual.name).toEqual(depositToken.name);
        expect(actual.symbol).toEqual(depositToken.symbol);
        expect(actual.pricePerFullShare).toEqual(snapshot.balance / snapshot.supply);
        const price = await strategy.getPrice(depositToken.address);
        expect(actual.value).toEqual(inCurrency(price, currency) * actual.balance);
      });
    };

    describe('non-digg token conversion', () => {
      testToSettBalance(TOKENS.BBADGER);
    });

    describe('digg token conversion', () => {
      testToSettBalance(TOKENS.BDIGG);
    });
  });

  describe('defaultBoost', () => {
    it('returns a boost with all fields as the default values', () => {
      const expected = {
        leaderboard: `${chain.network}_${LeaderBoardType.BadgerBoost}`,
        rank: 0,
        address: TEST_ADDR,
        boost: 1,
        stakeRatio: 0,
        nftBalance: 0,
        nativeBalance: 0,
        nonNativeBalance: 0,
      };
      expect(defaultBoost(chain, TEST_ADDR)).toMatchObject(expected);
    });
  });

  describe('getCachedBoost', () => {
    describe('no cached boost', () => {
      it('returns the default boost', async () => {
        setupMapper([]);
        const result = await getCachedBoost(chain, TEST_ADDR);
        expect(result).toMatchObject(defaultBoost(chain, TEST_ADDR));
      });
    });
    describe('a previously cached boost', () => {
      it('returns the default boost', async () => {
        const boost = defaultBoost(chain, TEST_ADDR);
        boost.rank = 42;
        boost.stakeRatio = 1;
        boost.nativeBalance = 32021;
        boost.nonNativeBalance = 32021;
        setupMapper([boost]);
        const result = await getCachedBoost(chain, TEST_ADDR);
        expect(result).toMatchObject(boost);
      });
    });
  });
});
