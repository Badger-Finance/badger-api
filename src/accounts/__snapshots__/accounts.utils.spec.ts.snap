// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts.utils getAccountMap a saved account returns the stored account 1`] = `
Object {
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Object {
    "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
    "balances": Array [],
    "boost": 0,
    "boostRank": 0,
    "claimableBalances": Array [],
    "earnedValue": 0,
    "multipliers": Array [],
    "nativeBalance": 0,
    "nonNativeBalance": 0,
    "stakeRatio": 0,
    "value": 0,
  },
}
`;

exports[`accounts.utils getAccountMap no saved account returns the account with the default account mapping 1`] = `
Object {
  "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484": Object {
    "address": "0xe6487033F5C8e2b4726AF54CA1449FEC18Bd1484",
    "balances": Array [],
    "boost": 0,
    "boostRank": 0,
    "claimableBalances": Array [],
    "earnedValue": 0,
    "multipliers": Array [],
    "nativeBalance": 0,
    "nonNativeBalance": 0,
    "stakeRatio": 0,
    "value": 0,
  },
}
`;
